File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.cpp
// Chain of Responsibility pattern in C++.
//
// Support ticket escalation: a virtual handler base class holds the
// next link; each level resolves tickets up to its competence or
// forwards them along the chain.

#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct Ticket {
    std::string subject;
    int severity; // 1 (trivial) .. 4 (critical)
};

class Handler {
public:
    virtual ~Handler() = default;

    Handler* setNext(std::unique_ptr<Handler> next) {
        next_ = std::move(next);
        return next_.get(); // enables a->setNext(b)->setNext(c)
    }

    virtual void handle(const Ticket& ticket) {
        if (next_) {
            next_->handle(ticket);
        } else {
            std::cout << "UNRESOLVED: \"" << ticket.subject << "\" (severity "
                      << ticket.severity << ") -> escalate to engineering\n";
        }
    }

private:
    std::unique_ptr<Handler> next_;
};

class SupportLevel : public Handler {
public:
    SupportLevel(std::string name, int maxSeverity)
        : name_(std::move(name)), maxSeverity_(maxSeverity) {}

    void handle(const Ticket& ticket) override {
        if (ticket.severity <= maxSeverity_) {
            std::cout << name_ << " resolved \"" << ticket.subject
                      << "\" (severity " << ticket.severity << ")\n";
        } else {
            Handler::handle(ticket);
        }
    }

private:
    std::string name_;
    int maxSeverity_;
};

int main() {
    auto chain = std::make_unique<SupportLevel>("Help bot", 1);
    chain->setNext(std::make_unique<SupportLevel>("Frontline agent", 2))
        ->setNext(std::make_unique<SupportLevel>("Support manager", 3));

    const std::vector<Ticket> tickets = {
        {"Reset my password", 1},
        {"Invoice shows wrong amount", 2},
        {"Account data corrupted", 3},
        {"Full region outage", 4},
    };
    for (const auto& ticket : tickets) {
        chain->handle(ticket);
    }
    return 0;
}

File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.py
"""Chain of Responsibility pattern in Python.

Support ticket escalation: each support level resolves tickets up to its
competence and passes the rest along the chain.

Note: for simple cases a list of callables tried in order is often more
Pythonic; the linked-handler form shown here pays off when chains are
assembled or reordered at runtime.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Ticket:
    subject: str
    severity: int  # 1 (trivial) .. 4 (critical)


class Handler:
    """Base handler: holds the next link and forwards unhandled tickets."""

    def __init__(self) -> None:
        self._next: Handler | None = None

    def set_next(self, handler: Handler) -> Handler:
        self._next = handler
        return handler  # enables a.set_next(b).set_next(c)

    def handle(self, ticket: Ticket) -> None:
        if self._next is not None:
            self._next.handle(ticket)
        else:
            print(f"UNRESOLVED: {ticket.subject!r} (severity {ticket.severity})"
                  " -> escalate to engineering")


class SupportLevel(Handler):
    def __init__(self, name: str, max_severity: int) -> None:
        super().__init__()
        self._name = name
        self._max_severity = max_severity

    def handle(self, ticket: Ticket) -> None:
        if ticket.severity <= self._max_severity:
            print(f"{self._name} resolved {ticket.subject!r} (severity {ticket.severity})")
        else:
            super().handle(ticket)


if __name__ == "__main__":
    chain = SupportLevel("Help bot", 1)
    chain.set_next(SupportLevel("Frontline agent", 2)).set_next(
        SupportLevel("Support manager", 3)
    )

    tickets = [
        Ticket("Reset my password", 1),
        Ticket("Invoice shows wrong amount", 2),
        Ticket("Account data corrupted", 3),
        Ticket("Full region outage", 4),
    ]
    for ticket in tickets:
        chain.handle(ticket)
